File: internal/compiler/compiler.go
package compiler

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/H1ghBre4k3r/go-bf/internal/input"
	"github.com/H1ghBre4k3r/go-bf/internal/lexer"
	"github.com/H1ghBre4k3r/go-bf/internal/parser"
)

type Compiler struct {
	path       string
	outputPath string
	program    string
}

func New(inputPath string, outputPath string) *Compiler {
	return &Compiler{
		path:       inputPath,
		outputPath: outputPath,
		program:    input.ReadFile(inputPath),
	}
}

func (c *Compiler) Start() {
	lexed := lexer.Lex(c.program, c.path)
	parsed := parser.Parse(lexed, c.path)
	compiled := c.compile(parsed)
	c.saveCode(fmt.Sprintf(base, compiled))
}

var label = 0

func (c *Compiler) compile(parsed []parser.Instruction) string {
	toReturn := ""
	for index := 0; index < len(parsed); index++ {
		i := parsed[index]

		switch i.Operation {
		case parser.MOVE:
			toReturn += fmt.Sprintf("; MOVE %v\n", i.Operand.(int))
			toReturn += fmt.Sprintf("    add	\trdi, %v\n", i.Operand.(int))

		case parser.ADD:
			toReturn += fmt.Sprintf("; ADD %v\n", i.Operand.(int))
			toReturn += fmt.Sprintf("	add 	byte[rdx+rdi], %v\n", i.Operand.(int))

		case parser.LOOP:
			jumpLabel := label
			label++
			toReturn += fmt.Sprintf("; LOOP to loop_%v\n", jumpLabel)
			toReturn += fmt.Sprintf("loop_%v_start: \n", jumpLabel)
			toReturn += "	cmp 	byte[rdx+rdi], 0\n"
			toReturn += fmt.Sprintf("	je		loop_%v_end\n", jumpLabel)

			toReturn += c.compile(i.Operand.([]parser.Instruction))

			toReturn += fmt.Sprintf("	jmp		loop_%v_start\n", jumpLabel)
			toReturn += fmt.Sprintf("loop_%v_end:\n", jumpLabel)

		case parser.PRINT:
			toReturn += "; PRINT\n"
			toReturn += "	lea 	rsi, [rdx+rdi]\n"
			toReturn += "	call 	print\n"

		case parser.READ:
			// maybe later
		}
	}
	return toReturn
}

func (c *Compiler) saveCode(code string) {
	outputPath := c.path + ".asm"
	if c.outputPath != "" {
		name := filepath.Base(c.path)
		outputPath = filepath.Join(c.outputPath, name+".asm")
	}
	if err := os.WriteFile(outputPath, []byte(code), 0644); err != nil {
		fmt.Println(err.Error())
		os.Exit(-1)
	}
}

File: src/compiler.rs
use std::{fmt::format, process};

use crate::{
    lexing::lexer::lex,
    parsing::{parser::parse, tokens::Token},
    util::read_file,
};

pub fn compile(filepath: &String) {
    let code = read_file(filepath);
    match code {
        Err(_) => {
            eprintln!("Could not open file '{}'!", filepath);
            process::exit(-1);
        }
        Ok(code) => {
            let lexed = lex(&code, filepath);
            let parsed = parse(&lexed, filepath);
            let compiled = _compile(parsed);
            println!(
                "
default rel
global _main

section .text

print:
    push    rax
    push    rdx
    push    rdi
    mov     rdi, 1 ; stdout
    mov     rax, 0x2000004 ; write
    mov     rdx, 1
    syscall
    pop     rdi
    pop     rdx
    pop     rax
    ret

_main: 
    mov     rdi, 0
    lea     rdx, array
{}
    mov     rax, 0x2000001 ; exit
    mov     rdi, 0 
    syscall

segment .bss
arraySize equ 3000
array: resb arraySize
            ",
                compiled
            )
        }
    }
}

fn _compile(parsed: Vec<Token>) -> String {
    let mut label = 1;
    let mut asm = String::from("");

    for instruction in parsed {
        match instruction {
            Token::MOVE(amount) => {
                asm += format!("; MOVE {}\n", amount).as_str();
                asm += format!("    add     \trdi, {}\n", amount).as_str();
            }
            Token::ADD(amount) => {
                asm += format!("; ADD {}\n", amount).as_str();
                asm += format!("    add     byte[rdx+rdi], {}\n", amount).as_str();
            }
            Token::LOOP(loop_content) => {
                let jump_label = label;
                label += 1;
                asm += format!("; LOOP for loop_{}\n", jump_label).as_str();
                asm += format!("loop_{}_start: \n", jump_label).as_str();
                asm += format!("    cmp     byte[rdx+rdi], 0\n").as_str();
                asm += format!("    je      loop_{}_end\n", jump_label).as_str();

                asm += _compile(loop_content).as_str();

                asm += format!("    jmp     loop_{}_start\n", jump_label).as_str();
                asm += format!("loop_{}_end: \n", jump_label).as_str();
            }
            Token::PRINT => {
                asm += format!("; PRINT\n").as_str();
                asm += format!("    lea     rsi, [rdx+rdi]\n").as_str();
                asm += format!("    call    print\n").as_str();
            }
            Token::READ => (),
        }
    }
    return asm;
}
